src/condor_daemon_client/dc_schedd.cpp

```diff
@@ -2104,3 +2104,277 @@ DCSchedd::requestImpersonationTokenAsync(const std::string &identity,
+	std::string projlist;
+	const char * projection = nullptr;
+
+	if ( ! attrs.empty()) {
+		for (auto attr : attrs) {
+			if (projlist.empty()) projlist += "\n";
+			projlist += attr;
+		}
+		send_server_time = attrs.count(ATTR_SERVER_TIME) > 0;
+		projection = projlist.c_str();
+	}
+
+	return makeUsersQueryAd(request_ad, constraint, projection, send_server_time, match_limit);
+}
+
+int DCSchedd::queryUsers(
+	const classad::ClassAd & query_ad,
+	// return 0 to take ownership of the ad, non-zero to allow the ad to be deleted after, -1 aborts the loop
+	int (*process_func)(void*, ClassAd *ad),
+	void * process_func_data,
+	int connect_timeout,
+	CondorError *errstack,
+	ClassAd ** psummary_ad)
+{
+	ClassAd *ad = NULL;	// user ad result
+
+	int cmd = QUERY_USERREC_ADS;
+
+	Sock* sock;
+	if (!(sock = startCommand(cmd, Stream::reli_sock, connect_timeout, errstack))) return Q_SCHEDD_COMMUNICATION_ERROR;
+
+	classad_shared_ptr<Sock> sock_sentry(sock);
+
+	if (!putClassAd(sock, query_ad) || !sock->end_of_message()) return Q_SCHEDD_COMMUNICATION_ERROR;
+	dprintf(D_FULLDEBUG, "Sent Users request classad to schedd\n");
+
+	int rval = 0;
+	do {
+		ad = new ClassAd();
+		if ( ! getClassAd(sock, *ad)) {
+			rval = Q_SCHEDD_COMMUNICATION_ERROR;
+			break;
+		}
+		dprintf(D_FULLDEBUG, "Got classad from schedd.\n");
+		std::string mytype;
+		if (ad->EvaluateAttrString(ATTR_MY_TYPE, mytype) && mytype == "Summary")
+		{ // Last ad.
+			dprintf(D_FULLDEBUG, "Ad was last one from schedd.\n");
+			std::string errorMsg;
+			int error_val;
+			if (ad->EvaluateAttrInt(ATTR_ERROR_CODE, error_val) && error_val && ad->EvaluateAttrString(ATTR_ERROR_STRING, errorMsg))
+			{
+				if (errstack) errstack->push("TOOL", error_val, errorMsg.c_str());
+				rval = Q_REMOTE_ERROR;
+			} else if ( ! sock->end_of_message()) {
+				rval = Q_SCHEDD_COMMUNICATION_ERROR;
+			}
+			sock->close();
+
+			if (psummary_ad && rval == 0) {
+				*psummary_ad = ad; // return the final ad, because it has summary information
+				ad = NULL; // so we don't delete it below.
+			}
+			break;
+		}
+
+		// Note: process_func() will return 0 if taking
+		// ownership of ad, so only delete if it returns non-zero, else set to NULL
+		// so we don't delete it here.  Either way, next set ad to NULL since either
+		// it has been deleted or will be deleted later by process_func().
+		// if process_func returns a negative value, we want to break out of the loop
+		int delit = process_func(process_func_data, ad);
+		if (delit) { delete ad; }
+		ad = NULL;
+		if (delit < 0) {
+			rval = -delit; // TODO: better error code here?
+			break;
+		}
+	} while (true);
+
+	// Make sure ad is not leaked no matter how we break out of the above loop.
+	delete ad;
+
+	return rval;
+}
+
+ClassAd* DCSchedd::actOnUsers (
+	int cmd, // ENABLE_USERREC or DISABLE_USERREC
+	const ClassAd * userads[], // either pass array of ad pointers
+	const char * usernames[], // or pass array of username pointers
+	int num_usernames,
```

Comment:
Splitting the array pointer and the size of the array it points to just gives people an opportunity to make a mistake, and it looks like you're generally converting to this representation from a `std::vector` anyway.  Consider switching to two `std::vector *`s here, instead?